experimental/incremental/executor.go

```diff
@@ -0,0 +1,153 @@
+// Copyright 2020-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package incremental
+
+import (
+	"context"
+	"runtime"
+	"slices"
+	"sync"
+
+	"golang.org/x/sync/semaphore"
+)
+
+// Executor is a caching executor for incremental queries.
+//
+// See [New], [Run], and [Invalidate].
+type Executor struct {
+	dirty sync.RWMutex
+	tasks sync.Map // [string, *task]
+
+	sema *semaphore.Weighted
+}
+
+// New constructs a new executor with the given maximum parallelism.
+//
+// Setting parallelism to zero or negative will default to GOMAXPROCS.
+func New(parallelism int) *Executor {
+	if parallelism <= 0 {
+		parallelism = runtime.GOMAXPROCS(0)
+	}
+
+	return &Executor{
+		sema: semaphore.NewWeighted(int64(parallelism)),
+	}
+}
+
+// Queries returns a snapshot of the URLs of which queries are present (and
+// memoized) in an Executor.
+//
+// The returned slice is sorted.
```

Comment:
It's kind of intended as a general debugging aid. I'm not sure we really need an iterator for this, because you really do want to sort them first and that requires buffering regardless.